### PR TITLE
Implementation of Tree View Protocol

### PR DESCRIPTION
**Pull request by danilbykov**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -151,14 +185,12 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
-    .then(classpath => {
-      launchMetals(
+    .then(() => {
+      return registerCommands(
         context,
-        javaPath,
-        classpath,
-        serverProperties,
-        javaOptions,
-        customRepositoriesEnv
```
Comment (ckipp01) [inline src/index.ts]: Why are we not including these anymore?

File: src/index.ts [hunk 2]
```diff
@@ -189,56 +221,12 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
-  javaPath: string,
-  metalsClasspath: string,
-  serverProperties: string[],
-  javaOptions: string[],
-  env: { COURSIER_REPOSITORIES?: string }
+  client: LanguageClient,
+  features: MetalsFeatures,
+  treeViewsManager: TreeViewsManager
 ) {
-  const baseProperties = [
```
Comment (ckipp01) [inline src/index.ts]: This is really important and needs to stay here. Why did you remove them?

File: src/index.ts [hunk 3]
```diff
@@ -189,56 +221,12 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
-    revealOutputChannelOn: RevealOutputChannelOn.Never
-  };
-
-  // TODO fix this any
-  // I know this any here is a bit gross, but I can't
-  // get any of the sendRequests to work or type check
-  // correctly without it. I notice that this was also
-  // done in some of the other coc-extensions, which
-  // game me the idea to try it this way, and it works
-  const client: any = new LanguageClient(
```
Comment (ckipp01) [inline src/index.ts]: Can you explain your reasoning for moving the client creation to up above instead of leaving it here?

File: src/index.ts [hunk 4]
```diff
@@ -189,56 +221,12 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
-    "metals",
-    "Metals",
-    serverOptions,
-    clientOptions
-  );
-
-  const features = new MetalsFeatures();
-  client.registerFeature(features);
```
Comment (ckipp01) [inline src/index.ts]: Again, going back to the features. This is ideally where we'd have the tvp feature registered, together with the others.

File: src/index.ts [hunk 5]
```diff
@@ -319,14 +305,11 @@ function launchMetals(
           const location =
             params.arguments && (params.arguments[0] as Location);
           if (location) {
-            const range = Range.create(
```
Comment (ckipp01) [inline src/index.ts]: The `metals-goto-location` stuff that you removed here will need to stay to answer those requests. I don't believe it's tree view specific?

File: src/index.ts [hunk 6]
```diff
@@ -319,14 +305,11 @@ function launchMetals(
...
-              location.range.start.character,
-              location.range.end.line,
-              location.range.end.character
-            );
-            workspace.jumpTo(location.uri, range.start);
-            workspace.selectRange(range);
+            await treeViewsManager.prepareWindowForGoto();
+            // It seems this line fixes weird issue with "returned a response with an unknown
+            // request id" after executing commands several times.
+            await wait(10);
```
Comment (ckipp01) [inline src/index.ts]: Why is this necessary?

**Comment by ckipp01** [comment]
@gabro pinging you just so that you're aware this pr exists. It's not really ready for a full review yet, but since it's quite large, you should probably be aware it exists.

File: src/index.ts [hunk 7]
```diff
@@ -43,13 +41,45 @@ import * as path from "path";
...
 
 export async function activate(context: ExtensionContext) {
   detechLauncConfigurationChanges();
   await checkServerVersion();
 
+  const mbServiceStat = services
+    .getServiceStats()
+    .find(stat => stat.languageIds.indexOf("scala") != -1);
+  if (mbServiceStat === undefined) return;
+  const metalsClient = services.getService(mbServiceStat.id).client;
```
[the remark this answers is not in the document]
Comment (danilbykov) [reply src/index.ts]: Maybe I use different configuration than you use. I have following lines in my file coc-settings.json:
```
{
  "languageserver": {
    "metals": {
      "command": "metals-vim",
      "rootPatterns": ["build.sbt"],
      "filetypes": ["scala", "sbt"]
    }
  }
}
```
So coc.nvim starts metals-vim client automatically. By some reason you start another second client which is not registered in coc.nvim at all. I don't understand why this client is needed.

File: src/index.ts (hunk 5 above)
Comment (danilbykov) [reply src/index.ts]: Metals sends `metals-goto-location` command when user opens selected class or trait from tree view 'metalsBuild'.

I don't like line
```
workspace.selectRange(range);
```
because it left vim in visual mode.

File: src/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ckipp01) [reply src/index.ts]: Registering the client in that manner doesn't make a ton of sense, since it sort of defeats the purpose of having the extension register the client. If you take a look at the coc-tsserver, coc-python, or really any of the other extensions, they all do it this way. It's done purposefully so the user doesn't have manually do it. For example, the goal is to simply have a user issue a `:CocInstall coc-metals`, and that does everything they need to get started. Again, this is how it works in the other coc extensions as well and how I think we should approach it. 

Have you by chance used `coc-metals` a bit before implementing this? If not, it may be a good idea to use it a bit to see how things work without having to manually register metals, and I think you'll get the idea of what that's done.

[the remark this answers is not in the document]
Comment (danilbykov) [reply src/index.ts]: OK, I removed most of destructive changes in index.ts file :) Hopefully it should work now in your environment.

File: src/MetalsFeatures.ts [hunk 8]
```diff
@@ -4,20 +4,17 @@ import {
 } from "vscode-languageserver-protocol";
 import { StaticFeature, workspace } from "coc.nvim";
 
-export interface TreeViewProvider {}
```
[the remark this answers is not in the document]
Comment (danilbykov) [reply src/MetalsFeatures.ts]: Because I defined `TreeViewFeature` in separate file tvp/feature.ts. File tvp/features.ts is quite big so I don't think we should define all features in one file. Debugging feature will probably require even more notifications than TreeViewProtocol.

File language-client/client.ts from coc.nvim also define a lot of features and they are defined in different classes. So I think we should follow the same pattern. What do you think?

File: package.json [hunk 9]
```diff
@@ -96,6 +97,98 @@
...
+          "description": "Force reloading children of this node. Maybe be useful when wrong result is cached and tree contains invalid data."
+        },
+        "treeviews.gotoParentNode": {
+          "type": "string",
+          "default": "p",
+          "description": "Go to parent node"
+        },
+        "treeviews.gotoFirstChild": {
+          "type": "string",
+          "default": "K",
```
[the remark this answers is not in the document]
Comment (ckipp01) [reply package.json]: Ahh, that makes sense. So actually playing around with it a bit now, going to `G` would be a terrible ideas since it'd hijack my ability to jump to the bottom. For the keys, I'd say I'll need to use this a bit for a while before figuring out what makes most sense. 👍

File: src/MetalsFeatures.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckipp01) [reply src/MetalsFeatures.ts]: Hmm, yea, I guess I'm open to it. Looking through it, I guess they don't absolutely need to be together, and if the VS Code extension it's easy to since there really isn't much else that goes into it, but seeing that this one is quite involved, maybe it makes sense. I also think seeing how this one is done, I could also move the decoration protocol one to be a bit more similar to this as well.

File: src/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ckipp01) [reply src/index.ts]: Haha thank you. I was able to pull this down and use it a bit. This is really incredible. Thanks a ton for your work. I'll keep playing around with things and give you some initial impressions and some more thoughts on the code.

This is exciting!

**Comment by ckipp01** [comment]
Hey @danilbykov thanks again for all the work on this! I'll be using it throughout the day today, and I'll start putting feedback in here as I come across it from a usability perspective. A couple initial thoughts.

After opening and closing the tvp panel a few times, you get leftover `[No Name]` buffers. We'll want to make sure those are cleared.
Screenshot 2020-01-20 at 08 48 01

In the `metalsCompile` window, I've never been able to get anything to appear under the `Ongoing compilations` section. What is the status of that? From a user perspective, when someone sees that a module is compile like in the image below, I'm assuming they are going to want to see it there as well, or it may cause confusion.

Screenshot 2020-01-20 at 08 51 23

Another thing that will need to be addressed is the ability for the panel to pick up the theme that the user has set. For example in my image below, not only does it bug me that it doesn't match my theme, but some of the colors are unreadable.

Screenshot 2020-01-20 at 08 56 14

Regarding the `t` when in the panel to execute the command and open a new tab, are you open to changing this to utilize a buffer workflow instead of a tab. I feel like we often misuse tabs when we should be using buffers, and for this situation I think it is better fitting to open this up as a new buffer.

Again, great work.

**Comment by ckipp01** [comment]
Another thing to think about is users that are using NerdTree. I don't personally use it, but I did download it to test with this. I know a lot of NerdTree users constantly have NerdTree open, and if possible, I'd like the interaction between NerdTree and the tvp panel to be similiar to the experience someone would have with VS Code. For example if I have NerdTree open and I open the tvp panel, it currently opens a bit oddly with half of the screen being the tvp panel. I've pasted an image below.

Screenshot 2020-01-20 at 09 13 35

I think I'd prefer that NerdTree gets closed first and then replaces by the tvp protocol. Thoughts?

**Comment by danilbykov** [comment]
> After opening and closing the tvp panel a few times, you get leftover [No Name] buffers. We'll want to make sure those are cleared.

Looks like buffer leak. Will investigate.

> In the metalsCompile window, I've never been able to get anything to appear under the Ongoing compilations section. What is the status of that?

It works in my environment and shows lines like "ProjectName - X seconds (N %)" during compilation. Could you please send me file ~/.cache/metals/lsp.trace.json when this panel is active and you recompiled something? I will try narrow down problem.

> Another thing that will need to be addressed is the ability for the panel to pick up the theme that the user has set.

Will check.

> Regarding the t when in the panel to execute the command and open a new tab, are you open to changing this to utilize a buffer workflow instead of a tab.

Have you seen that I also have hotkey "i" to open in horizontal split and "s" to open in vertical split? Though maybe I don't understand what you mean by "utilize a buffer workflow instead of a tab".

> I think I'd prefer that NerdTree gets closed first and then replaces by the tvp panel. Thoughts?

Good idea. Another option is to open TreeViews on right side. Probably need to add special settings for this and let user to configure.

**Comment by ckipp01** [comment]
> > In the metalsCompile window, I've never been able to get anything to appear under the Ongoing compilations section. What is the status of that?
> 
> It works in my environment and shows lines like "ProjectName - X seconds (N %)" during compilation. Could you please send me file ~/.cache/metals/lsp.trace.json when this panel is active and you recompiled something? I will try narrow down problem.

Alright, so playing around with this a bit I _am_ able to get it to appear every once in a while, but not as much as I'd assume, and it seems that it's a bit spotty? For example in the screen recording there seems to be messages for compile mtags and interfaces, but those never appear in the tvp panel. I've included a screen recording below.

https://i.imgur.com/VP1Dvfz.mp4

And I have the logs for this session here.

https://gist.github.com/ckipp01/ecc91024bfe16cb74dee7f101f1e5aa6

Hopefully that will help a bit. Also notice that at times the MetalsCompile window shifts down a bit? I'm not sure why that happens? You can notice that it's taller towards the end than in the beginning.

> > Regarding the t when in the panel to execute the command and open a new tab, are you open to changing this to utilize a buffer workflow instead of a tab.
> 
> Have you seen that I also have hotkey "i" to open in horizontal split and "s" to open in vertical split? Though maybe I don't understand what you mean by "utilize a buffer workflow instead of a tab".

Ahh I totally missed the `i` and `s`. Nice. Would it be possible to do `v` for the vertical split and `s` for split?

That makes my point about the buffer vs tabs pointless, so ignore it.

**Comment by danilbykov** [comment]
Regarding metalsCompile treeview. It looks like it is Metals issue.

Vim receives information about available treeviews:
```
[Trace - 02:17:33 PM] Sending notification 'metals/treeViewDidChange'
Params: {
  "nodes": [
    {
      "viewId": "metalsBuild"
    },
    {
      "viewId": "metalsCompile"
    }
  ]
}
```
and tells to Metals that treeview metalsCompile is opened
```
[Trace - 02:17:36 PM] Received notification 'metals/treeViewVisibilityDidChange'
Params: {
  "viewId": "metalsCompile",
  "visible": true
}
```
and opens "Ongoing Compilations" node automatically:
```
[Trace - 02:17:37 PM] Received notification 'metals/treeViewNodeCollapseDidChange'
Params: {
  "viewId": "metalsCompile",
  "nodeUri": "metals://ongoing-compilations",
  "collapsed": false
}
[Trace - 02:17:37 PM] Received request 'metals/treeViewChildren - (5)'
Params: {
  "viewId": "metalsCompile",
  "nodeUri": "metals://ongoing-compilations"
}
[Trace - 02:17:37 PM] Sending response 'metals/treeViewChildren - (5)'. Processing request took 1ms
Result: {
  "nodes": []
}
```
In last response we see empty response which means there is no ongoing compilations.

Then vim waits for notification 'metals/treeViewDidChange' for node "ongoing-compilations" and receives only at 02:18:09 PM:
```
[Trace - 02:18:09 PM] Sending notification 'metals/treeViewDidChange'
Params: {
  "nodes": [
    {
      "viewId": "metalsCompile",
      "nodeUri": "metals://ongoing-compilations",
      "label": "Ongoing compilations (1)",
      "collapseState": "collapsed"
    }
  ]
}
```
This notification tells vim to rebuild children of this node so vim requests updated children:
```
[Trace - 02:18:09 PM] Received request 'metals/treeViewChildren - (22)'
Params: {
  "viewId": "metalsCompile",
  "nodeUri": "metals://ongoing-compilations"
}
[Trace - 02:18:09 PM] Sending response 'metals/treeViewChildren - (22)'. Processing request took 1ms
Result: {
  "nodes": [
    {
      "viewId": "metalsCompile",
      "nodeUri": "file:/Users/ckipp/Documents/scala-workspace/metals/metals/?id\u003dmetals",
      "label": "metals - 0s (10%)"
    }
  ]
}
```
and gets "metals - 0s (10%)". There is no calls with 'metals/treeViewDidChange' between 02:17:37 and 02:18:09 so treeview was not updated. It works as described in protocol.

**Comment by ckipp01** [comment]
Sounds good @danilbykov, thanks for diving into that. I'll pay a bit more attention to it see if I can track down what's happening. Honestly, I think it's actually more of an issue with a message not clearing. I'll look into that to see if we can figure out a way to potentially do that.

The buffer leakage looks fixed indeed. Great job!

I'll start diving into the code now since everything seems to be working well. The other two things that we still need to look at are

[ ] Making sure the theme is picked up from the user
[ ] If NerdTree is open, close it when you open the tvp panel.

File: package.json [hunk 10]
```diff
@@ -96,6 +97,98 @@
...
+              "size": 40,
+              "expanded": [
+                [
+                  "projects:"
+                ]
+              ]
+            }
+          ]
+        },
+        "treeviews.toggleNode": {
```
Comment (ckipp01) [inline package.json]: Are we able to prepend metals to all of these to stay consistent with the others. For example `metals.treeviews.toggleNode`.

File: src/tvp/treeviews.ts [hunk 11]
```diff
@@ -291,7 +299,33 @@ export class TreeViewsManager implements Disposable {
...
+    if (restoreNerdTree) {
+      await tabpage.setVar("TVP_restoreNerdTree", false, false)
+      await this.nvim.eval("NERDTreeCreator.ToggleTabTree('')")
+    }
+  }
+
+  private async getPanelAlignment(): Promise<string> {
+    if (this.config.get<string>("NERDTreeIntegration") === "OpenTreeViewsOnOppositeSide"
+         && await this.nvim.eval("NERDTree.IsOpen()") === 1
+         && await this.nvim.getVvar("NERDTreeWinPos") === "left") {
```
Comment (ckipp01) [inline src/tvp/treeviews.ts]: This should be `getVar`

File: package.json (hunk 10 above)
Comment (danilbykov) [reply package.json]: Sure. Will do soon.

**Comment by ckipp01** [comment]
@danilbykov really great job on all of this. I'm curious to get your thoughts on something. So in conversation with a couple others the question came up if we should have something as specific as NERDTree integration, since it's very specific to that plugin. Also, if the user is using a different file viewer plugin like `coc-explorer`, this won't help. Do you think it's trying to do too much by detecting NERDTree? I know it's not a ton of code, but I sort of dislike having NERDTree specific stuff in here. Do you think by just having the ability to set the panel to the opposite side would suffice? I'm tempted to have you hold off on the NERDTree commit that you just pushed, have you save it, and then have more discussion about it. Thoughts?

The theme integration looks great btw.

**Comment by ckipp01** [comment]
Another reason for me saying that now is that I realize this PR is huge. If possibly, I'd like to solidify what we have now, use it for a bit, and then address issues that come up rather than trying to ensure that we have everything perfect in this PR.

File: package.json (hunk 10 above)
Comment (danilbykov) [reply package.json]: Done.

**Comment by danilbykov** [comment]
I also was using this plugin for about 2 or 3 last weeks and I can say it has drawbacks. For example:
1. it is enough slow (in comparison with plain NERDtree)
2. it can be started only after full initialization of metals and after end of compilation.
3. it stops working if project has compilation error. (Even more. Compilation error leads to hung up. Metals sends notification that root node is updated. Coc-metals tries to reload information about all open nodes. Then Metals again sends notification that root node is updated and process repeats infinitely. It is bug of metals and can reproduced in VS code too. The only way to stop it is closing treeview panel.)
4. it doesn't allow create/delete files.
5. it is redrawn each time when code is changed and project is recompiled.

The only advantage of this plugin is that it allows to review list of classes from project libraries (jar files). It is not available with NERDTree and very useful to learn code in libraries.

Some drawbacks can be fixed. For example.

Issue in 3rd item must be fixed in Metals.

Issue in 4th item can be improved. Metals can send information about file where class/trait/object is defined. I believe it is available. Then we can add hotkeys to delete current file or create new file in the same package (directory) as current.

Issue in 5th item also can be improved. I can compare initial and updated trees and do not update view if they are the same.

But issue from first and second item can not be fixed. So I think that treeview plugin must be used as complement for NERDTree (or another explorer plugin). I think it plays the same role in VScode. So we should provide even more smooth integration with NERDTree. For example, allow to switch between treeviews and NERDTree and synchronize open packages/directories.

**Comment by ckipp01** [comment]
Hey @danilbykov, thanks for all the input. It's helpful.

Regarding the points you brought up:

1. Yea, true, but that's sort of expected a bit no? It's doing quite a bit more than just listing the files. I do notice a slight lag in some of the larger projects I've used it in, but not enough for it to bug me.
2. Just acknowledging that yea, it's a drawback
3. Are you able to submit an issue for this in the [main metals issues](https://github.com/scalameta/metals/issues). I'm not sure how we'll address it, but I think that it's good to have a conversation about it.
4. True, but that's also not really the purpose of the protocol
5. I think that would be a great approach.

I get what you're saying about it being a companion of sorts for NERDTree, and yea, honestly I'm going to guess that NERDTree has at least 75% of the people that may be using coc-metals. I find myself in the 25% that don't as I'm a heavy netrw user. I guess like you said, have a smooth integration is key.

My proposal would be this: Let's indeed make sure that this works smoothly with NERDTree, but let's not include it in this PR. I'm not a fan of giant pr's as they become really hard to review. I think we should solidify up the core, merge if we are happy with it, and then start another pr with the NERDTree related stuff. Thoughts?

File: package.json [hunk 12]
```diff
@@ -156,6 +254,6 @@
     "metals-languageclient": "^0.1.9",
     "promisify-child-process": "^3.1.3",
     "semver": "^7.1.1",
-    "vscode-languageserver-protocol": "^3.15.1"
+    "vscode-languageserver-protocol": "3.15.0-next.6"
```
Comment (danilbykov) [inline package.json]: Do you have compilation error in file index.ts in lines with `client.onRequest`, `client.sendNotification`? This change fixes error. I took this version from coc.nvim.

**Comment by danilbykov** [comment]
> My proposal would be this: Let's indeed make sure that this works smoothly with NERDTree, but let's not include it in this PR.

Sounds good.

File: src/tvp/controller.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    this.listeners.push(
+      treeViewFeature.providerEvents()(provider => {
+        const model = new TreeModel(provider)
+        this.treeModels.set(provider.viewId, model)
+        treeViewsManager.addTreeModel(model)
+      })
+    )
+
+    this.listeners.push(
+      commands.registerCommand("metals.tvp", viewName => {
```
Comment (ckipp01) [inline src/tvp/controller.ts]: Can we be more explicit with `viewName` here, especially since this will only ever be `metalsBuild` or `metalsCompile`. Maybe an Enum?

File: src/tvp/controller.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        if (viewName === undefined) {
+          treeViewsManager.toggleAllTrees()
+        } else {
+          treeViewsManager.toggleTreeView(viewName)
+        }
+      })
+    )
+
+    this.listeners.push(
+      commands.registerCommand("metals.tvp.view", (action, viewId) => {
```
Comment (ckipp01) [inline src/tvp/controller.ts]: So this is also showing up in the command pallete with no description. Can we add it into the `package.json` with a title so it will show when a user sees it in the command pallete? Same as with `metals.tvp`, can we extract action into a type?

File: src/tvp/controller.ts [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+          case "Hidden":
+            return treeViewsManager.viewHidden(viewId)
+          default:
+            return
+        }
+      })
+    )
+
+    this.listeners.push(
+      commands.registerCommand("metals.revealInTreeView", async viewName => {
```
Comment (ckipp01) [inline src/tvp/controller.ts]: Same, can we give `viewName` a type.

File: src/tvp/feature.ts [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+} from 'vscode-languageserver-protocol'
+import {
+  TreeViewChildrenParams, TreeViewChildrenResult, TreeViewDidChangeParams,
+  TreeViewNodeCollapseDidChangeParams, TreeViewVisibilityDidChangeParams, TreeViewNode,
+  MetalsTreeRevealResult
+} from './domain'
+import { commands } from 'coc.nvim'
+import { TreeViewProvider } from './provider'
+
+interface TreeViewParams { }
```
Comment (ckipp01) [inline src/tvp/feature.ts]: Are we able to complete these interfaces?

File: src/tvp/feature.ts [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+interface TreeViewOptions { }
```
Comment (ckipp01) [inline src/tvp/feature.ts]: Same here, can we fill this out.

File: src/tvp/feature.ts [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+            viewId,
+
+            updatedNodes: updatesEmitter.event,
+
+            loadNodeChildren: (
+              parentNode?: string
+            ): Promise<TreeViewNode[]> => {
+              const result = client
+                .sendRequest(this.treeViewChildrenParamsType, { viewId, nodeUri: parentNode })
+                .then(r => r.nodes)
```
Comment (ckipp01) [inline src/tvp/feature.ts]: Nit: can we just use `response` rather than `r`

File: src/tvp/feature.ts [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+  public initialize(
+    capabilities: ServerCapabilities,
+  ): void {
+    if (!capabilities.experimental!.treeViewProvider) return
+    const client = this._client
+
+    client.onNotification(this.treeViewDidChangeType, message => {
+      message.nodes.forEach(node => {
+        const viewId = node.viewId
+        const mbViewUpdater = this.viewUpdaters.get(viewId)
```
Comment (ckipp01) [inline src/tvp/feature.ts]: Can we change the name of this to something else? It's not super obvious what the `mb` here is referring to, and I think it'll be much clearer if we changed it.

File: src/tvp/model.ts [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+      return false
+    }
+  }
+
+  public async refreshSubtree(viewNode: TreeViewNode | undefined): Promise<void> {
+    async function findOpenNodes(acc: string[][], curPath: string[], node: Node): Promise<string[][]> {
+      if (node.isExpanded()) {
+        acc.push(curPath)
+        const children = await node.getChildren()
+        return children.reduce((zP, child) => {
```
Comment (ckipp01) [inline src/tvp/model.ts]: Can we use something more explicit that `zP`?

File: src/tvp/model.ts [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+        return acc
+      }
+    }
+
+    async function reloadSubtree(node: Node, openNodes: string[][]): Promise<void> {
+      if (openNodes.length != 0) {
+        const children = await node.getChildren()
+        const childToNodes = groupBy(openNodes.filter(node => node.length != 0), arr => arr[0])
+        await Promise.all(children.map(child => {
+          const mbOpenNodes = childToNodes.get(child.viewNode.nodeUri)
```
Comment (ckipp01) [inline src/tvp/model.ts]: Same comment as above. I think for future maintainability, changing mb will be easier for a ready to quickly skim and see what it's referring to.

File: src/tvp/model.ts [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+      return 1
+    }
+  }
+
+  private async collectVisibleNodes(): Promise<NodeView[]> {
+    async function collectInternal(acc: NodeView[], node: Node): Promise<NodeView[]> {
+      const newAcc = acc.concat(node.makeView())
+      if (node.isExpanded()) {
+        const children = await node.getChildren()
+        return children.reduce((zP, child) => zP.then(z => collectInternal(z, child)), Promise.resolve(newAcc))
```
Comment (ckipp01) [inline src/tvp/model.ts]: Can we change `zP` here?

File: src/tvp/model.ts [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+    }
+  }
+
+  private async findNodeWithOffsetInternal(node: Node, offset: number): Promise<Node | undefined> {
+    if (offset == 0) {
+      return node
+    } else {
+      const children = await node.getChildren()
+      return children
+        .reduce((stateP, child) => {
```
Comment (ckipp01) [inline src/tvp/model.ts]: Can we change the variable name `stateP`? Not a huge fan of ending with a cap letter, and also a fan of more explicit naming even if it's a bit more verbose.

File: src/tvp/model.ts [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+
+  public async findNodeOffset(nodeView: NodeView): Promise<number | undefined> {
+    async function findNodeOffsetInternal(acc: [number, boolean], node: Node): Promise<[number, boolean]> {
+      if (node.viewNode.nodeUri === nodeView.underlying.nodeUri || acc[1]) {
+        return [acc[0], true]
+      } else if (node.isExpanded()) {
+        const children = await node.getChildren()
+        const init: [number, boolean] = [acc[0] + 1, acc[1]]
+        return children.reduce((zP, child) => {
+          return zP.then(z => findNodeOffsetInternal(z, child))
```
Comment (ckipp01) [inline src/tvp/model.ts]: Same `zP` comment as before.

File: src/tvp/model.ts [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+    })
+  }
+
+  public async findParentNode(node: Node): Promise<Node | undefined> {
+    async function internal(result: Node | undefined, parent: Node): Promise<Node | undefined> {
+      if (result !== undefined) {
+        return result
+      } else if (parent.isExpanded()) {
+        const children = await parent.getChildren()
+        return children.reduce((zP, child) => {
```
Comment (ckipp01) [inline src/tvp/model.ts]: Same `zP` comment.

File: src/tvp/treeview.ts [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+  public get bufferId(): number {
+    return this.buffer.id
+  }
+
+  private async method(window: Window, topLine: number, curLine: number): Promise<void> {
+    await window.setCursor([topLine, 1])
+    await window.setCursor([curLine, 1])
+  }
+
+  public handleModelUpdates(ev0: TreeModelUpdate): void {
```
Comment (ckipp01) [inline src/tvp/treeview.ts]: Not a big fan of `ev0`. Can we change this to something more explicit?

**Comment by ckipp01** [comment]
@gabro, if you have any free time, I'd love another set of eyes to review this. Especially from someone that knows TS better than myself.

File: package.json (hunk 12 above)
Comment (ckipp01) [reply package.json]: hmmm, no, I haven't had any issues with it.

File: src/tvp/treeview.ts (hunk 26 above)
Comment (danilbykov) [reply src/tvp/treeview.ts]: Done.

File: src/tvp/feature.ts (hunk 17 above)
Comment (danilbykov) [reply src/tvp/feature.ts]: Removed

File: src/tvp/feature.ts (hunk 16 above)
Comment (danilbykov) [reply src/tvp/feature.ts]: Removed

File: src/tvp/controller.ts (hunk 14 above)
Comment (danilbykov) [reply src/tvp/controller.ts]: Added.

File: src/tvp/controller.ts (hunk 13 above)
Comment (danilbykov) [reply src/tvp/controller.ts]: First of all let me say that code inside directory src/tvp doesn't depend on `metalsBuild` or `metalsCompile`. Metals can add new views or remove existing one but it should work without any changes in coc-metals. User should just update hotkeys in .vimrc.

I don't understand what we can get here by adding enum.

File: src/tvp/feature.ts (hunk 18 above)
Comment (danilbykov) [reply src/tvp/feature.ts]: Done.

File: src/tvp/model.ts (hunk 20 above)
Comment (danilbykov) [reply src/tvp/model.ts]: I think that such names are quite natural for these variables in scala world.

`z` is standard name for accumulator of `foldLeft`, `foldRight` or `reduce`. Examples are [foldRight](https://github.com/scala/scala/blob/v2.13.1/src/library/scala/collection/immutable/List.scala#L320) or [foldLeft](https://github.com/scala/scala/blob/v2.12.10/src/library/scala/collection/LinearSeqOptimized.scala#L122).

Last capital letter is usually effect which is used somehow in current function or variables. This is used, for example, in [Monix](https://monix.io/api/3.0/monix/reactive/Observable.html) where a lot of methods have such naming: findL, firstL, completedL, comletedF, bracketF and so on. Cats-effect use [the same](https://typelevel.org/cats-effect/api/cats/effect/Concurrent.html).

So `zP` just means accumulator of function `reduce` wrapped inside `Promise`.

If you are disagree then I will rename.

File: src/tvp/model.ts (hunk 21 above)
Comment (danilbykov) [reply src/tvp/model.ts]: `mb` means maybe. Type of this variable `string | undefined` and this name helps to remember that we should check variable for `null`. There is even special option in typescript compiler which requires to check null-reference explicitly but this option is not active. So this name at least helps to remember about possible `null` value.

File: package.json (hunk 12 above)
Comment (danilbykov) [reply package.json]: I believe this is why you declared `client` as `any`:
https://github.com/scalameta/coc-metals/blob/master/src/index.ts#L218-L224

File: src/tvp/controller.ts (hunk 13 above)
Comment (ckipp01) [reply src/tvp/controller.ts]: Ah, I misunderstood then. Sounds good.

File: src/tvp/model.ts (hunk 20 above)
Comment (ckipp01) [reply src/tvp/model.ts]: Understood. Thanks for the explanation. I'm fine with leaving them then.

File: src/tvp/model.ts (hunk 21 above)
Comment (ckipp01) [reply src/tvp/model.ts]: Alright. Sounds good. Thanks for the explanation.

**Comment by gabro** [comment]
@ckipp01 I'll try to take a look, but don't feel blocked by me. I've had little time these days and I've prioritized some work on metals-languageclient so far.

Cool job @danilbykov by the way :)

**Comment by ckipp01** [comment]
Sounds good @gabro! I feel confident with this, so let's get it merged in, get some users, and let's see what issues we need to address. I don't plan on releasing a stable for a bit until this gets a bit more usage.

Again, great job @danilbykov 🎉

**Comment by danilbykov** [comment]
Thanks!
